<center> <h2> DS 3000 - Fall 2020</h2> </center>
<center> <h3> DS Report </h3> </center>

<center> <h3> Genre Predictions of songs based on Spotify Attributes</h3> </center>
<center><h4>Benjamin Wolff, Edwin Shen, Tatum Gaudinski Whitehead</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:

Music can be very hard to classify as genres may seem arbitrary. For instance, what makes a song rock? Pop? Jazz? An artist may say that their song is one genre while the public thinks it is another. Is one group correct? In this project, we strive to answer these questions with quantitative data. We examined a playlist of thousands of songs pulled from Spotify and used k-nearest neighbors, support vector machine, and decision tree algorithms to predict song genres. After fitting these models to our training data and performing hyperparameter tuning, our results showed that songs could be accurately classified using their features. Our decision tree model resulted in the best performance, and our work opens up opportunities for future projects involving the music genre classification.




<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION

Problem Statement
* Songs can be hard to classify into genres because the process is quite subjective. As a result, there is often disagreement and dispute about which genre a song may lie within. Thus, we would like to be able to accurately label songs based on their attributes. These include but are not limited to Beats per Minute, Danceability, Energy, and Liveness. In essence, we seek to identify if there are quantitative features that can classify song genres, and if so, be able to accurately classify any given song’s genre.

Significance of the Problem
* This is an important problem to tackle because there is limited data about music genre classification. After all, there were no other public/open-source projects we found that attempted genre classification with Spotify attributes. However, our project could provide meaningful insight into and new understandings about song-genre classifications. This knowledge could be used in the music industry to better classify songs, optimize recommended songs based on those classifications, and increase music application usage time and revenue. These insights could also help people better understand the music they listen to and aid them in finding their next favorite song. 

* One study attempted, with moderate success, to predict which songs will become big hits given its Spotify song attributes. Researchers made predictions using a Logistic Regression Model, a Neural Network, a SVM Model, and a Random Forest Decision Tree model. The validation accuracy of each model's predictions are as follows:
Logistic Regression: 81%,
Neural Network: 82%,
SVM: 89%,
Random Forest: 83%. (1)

* Another study attempted to predict genre labels given by the
Billboard charts using only chords and lyrics. The dataset used consisted of 190,165 musical segments from 5,305 pop songs. Researchers used a k-NN classifier and achieved an accuracy of 37.9% when the model was trained using both song chords and song lyrics. (2)

* Finally, one study attempted to identify song genres based on the song's waveform, where ML algorithms looked for patterns in wave graphs (spectrograms). The spectrograms were generated from timeslices of songs and then input into a neural network. The neural network had a total of 5 layers and achieved a 67% accuracy on testing data. Overall, this study was to see how well ML algorithms can classify music if they are given short song segments. (3)


Questions
<br> Given the above, we set out to answer the following questions:
*   Can we accurately classify a song's genre given it's Spotify attributes?
*   Are certain song attributes more indicative of the songs' genre than other attributes? If so, which ones are the most important? Are there any relationships between these attributes?
*   Will the KNN and SVM algorithms outperform the decision-tree algorithm when predicting a song’s genre based on the feature variables? If so, how much better or worse does the KNN algorithm perform compared to the SVM algorithm?

Hypotheses
<br>About the models:<br>
$H_{0}$: The decision tree model will not perform better (in terms of accuracy) than the KNN and SVM classification models when classifying song genres. <br>
$H_{1}$: The decision tree model will perform better than the KNN and SVM classification models when classifying song genres.

About the features:<br>
$H_{0}$:There is not a strong correlation between a song's danceability and energy. <br>
$H_{1}$:There is a strong correlation between a song's danceability and energy.

Note: Our alpha value will be .05.




<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

We obtained our data from a dataset on Kaggle (4). This dataset contains 41,099 songs from various genres and has 20 columns about the songs' attributes. Columns include data about track name, artist name, and more. Songs in the dataset date back as far as 1960 and go up to 2010. As this data was created by people on Kaggle, no web scraping was required to obtain it. For convenience, we stored the dataset on GitHub here: https://raw.githubusercontent.com/edshen17/FP4/master/genre_music.csv.


Below are the column names and descriptions. Please note that most of these attributes are directly from the Spotify API (5).
*   **track** - track name
*   **artist** -  artist name
*   **danceability** - how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.
*   **energy** - measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy.
*   **key** - estimated overall key of the track. Integers map to pitches using standard Pitch Class notation.
*   **loudness** - loudness of a track in decibels (dB). Values range between -60 and 0 db.
*   **mode** - modality (major or minor) of a track, the type of scale from which its melodic content is derived. Major is represented by 1 and minor is 0.
*   **speechiness** - detection of the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value.
*   **acousticness** - confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.
*   **instrumentalness** - prediction of whether a track contains no vocals.  The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.
*   **liveness** - detection of the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live. A value above 0.8 provides strong likelihood that the track is live.
*   **valence** - measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).
*   **tempo** - overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.
*   **duration in seconds** - duration of the track in seconds.
*   **time signature** - estimated overall time signature of a track. The time signature (meter) is a notational convention to specify how many beats are in each bar (or measure).
*   **decade** - decade the song was created in.
*   **genre** - the genre/category a song belongs to.
*   **chorus_hit** - likeliness of chorus being a hit (%)
*   **sections** - how many sections the song has
*   **popularity** - number of times the song has been on a popularity list


In [ ]:
import pandas as pd
import numpy as np
df_original = pd.read_csv('https://raw.githubusercontent.com/edshen17/FP4/master/genre_music.csv')
df_original

track            artist  ...  decade  genre
0      Jealous Kind Of Fella     Garland Green  ...     60s    edm
1              Initials B.B.  Serge Gainsbourg  ...     60s    pop
2               Melody Twist       Lord Melody  ...     60s    pop
3              Mi Bomba Sonó        Celia Cruz  ...     60s    pop
4                Uravu Solla       P. Susheela  ...     60s    r&b
...                      ...               ...  ...     ...    ...
41094          Lotus Flowers             Yolta  ...     10s   rock
41095      Calling My Spirit       Kodak Black  ...     10s    pop
41096          Teenage Dream        Katy Perry  ...     10s    pop
41097         Stormy Weather    Oscar Peterson  ...     10s    pop
41098                   Dust       Hans Zimmer  ...     10s   rock

[41099 rows x 20 columns]

### 2.2. Data Analysis


### &emsp; &emsp; 2.2.1. Variables

1.   The independent variable for the first hypothesis (about the models) is the classification algorithm used (Decision tree/KNN/SVM). The dependent variable is the test set accuracy score on these models.

2.   The independent variables for the second hypothesis (about the features) are the danceability and energy of a given song. The dependent variable is the calculated t-statistic.







### &emsp; &emsp; 2.2.3. Model/Algorithms

For this data, we are using classification models to classify the songs by genre. The target variable is the genre and the feature variables are the attributes of the songs provided by the Spotify API/our dataset. More specifically, the core feature variables we use are: danceability, energy, speechiness, tempo, duration_s (feature selection from section 3.1.4). 

This is a supervised learning classification problem. The size of our data allows us to split it into training and testing sets.

The algorithms we plan to use:
*   **K-Nearest Neighbor** - Compares inputted data points to the k data points that are "closest" to the sample and determines the class based on those points (most frequent class is chosen). We chose this algorithm because with the large amount of data we have, comparing an input to similar data points should provide an accurate and reasonable classification. Also, simplicity makes this algorithm a solid choice.
*   **Support Vector Machine (SVM)** - Looks to create "boundaries" that maximize the differences between the given classes and their data points. More specifically, it tries to find the hyperplane/decision surface that maximizes the margin between classes. We chose to use this algorithm because it is very effective in high dimensional spaces such as the one we are working with.

*   **Decision Tree** - Constructs a tree that asks yes/no questions regarding the feature values. We chose this algorithm because it is fairly simple, effective, and could be visualized if necessary. The disadvantage, however, is that it may not be able to understand some more complex concepts, which we will keep in mind when exploring this algorithm.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import LinearSVC
from sklearn.tree import DecisionTreeClassifier

# Creating the objects with the algorithms to be used later
estimators = { 
    'k-Nearest Neighbor': KNeighborsClassifier(),
    'Support Vector Machine': LinearSVC(max_iter=100000, random_state=3000),
    'Decision Tree': DecisionTreeClassifier(random_state=3000),
}

<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

### 3.1. Data Wrangling

### &emsp;3.1.0 Data Cleaning
Although our dataset comes from Kaggle, there are some columns that are unneeded for our analysis (eg. song popularity, artists, sections, etc.). Thus, we perform some data cleaning to make sure that there are only relevant feature/target variables present. Also, because there are so many elements in the dataset (40k+ songs), there may be some invalid or duplicate values. We have to ensure that those elements are dropped completely to avoid any problems.


In [ ]:
#drop unneeded columns, NA columns, and duplicates (cleaning). We also rename the "genre" column to "target"
df = df_original.drop(columns=['track', 'artist', 'chorus_hit', 'sections', 'popularity']).dropna().drop_duplicates().rename(columns={'genre': 'target'})
df.head()

danceability  energy  key  ...  time_signature  decade  target
0         0.417   0.620    3  ...               3     60s     edm
1         0.498   0.505    3  ...               4     60s     pop
2         0.657   0.649    5  ...               4     60s     pop
3         0.590   0.545    7  ...               4     60s     pop
4         0.515   0.765   11  ...               4     60s     r&b

[5 rows x 15 columns]

### &emsp;3.1.1 Data Wrangling

There are also some columns that need to be reformated. For example, the column "decade" gives us the decade the song was made in, but it does so in a string format (eg. "60s"). However, this column may be more valuable for us if we change it to contain int values. Similarly, the target variable (genre) contains string variables and hence cannot be used by scikitlearn. We would need to transform this categorical variable into a quantitative one so our ML algorithm can understand it.

In [ ]:
#function used to format the decade column (data wrangling)
def format_decade(decade):
    return int(decade[0:2])
df['decade'] = df['decade'].apply(format_decade)

#create dictionaries to hold values of the genres
genre_dict = {1: "pop", 2: "r&b", 3: "rock", 4: "latin", 5: "rap", 6: "edm"}
genre_nominal = {"pop": 1, "r&b": 2, "rock": 3, "latin": 4, "rap": 5, "edm": 6}

#to transform the genres into numbers
def transform_genre_name(column):
    return genre_nominal[column]

#create variables to store feature and target columns
df["target"] = df["target"].apply(transform_genre_name)
target = df['target']
features = df.loc[:, 'danceability':'decade']
df.head()

danceability  energy  key  ...  time_signature  decade  target
0         0.417   0.620    3  ...               3      60       6
1         0.498   0.505    3  ...               4      60       1
2         0.657   0.649    5  ...               4      60       1
3         0.590   0.545    7  ...               4      60       1
4         0.515   0.765   11  ...               4      60       2

[5 rows x 15 columns]

In [ ]:
target

0        6
1        1
2        1
3        1
4        2
        ..
41094    3
41095    1
41096    1
41097    1
41098    3
Name: target, Length: 40750, dtype: int64

In [ ]:
features

danceability  energy  key  ...  duration_s  time_signature  decade
0             0.417   0.620    3  ...     173.533               3      60
1             0.498   0.505    3  ...     213.613               4      60
2             0.657   0.649    5  ...     223.960               4      60
3             0.590   0.545    7  ...     157.907               4      60
4             0.515   0.765   11  ...     245.600               4      60
...             ...     ...  ...  ...         ...             ...     ...
41094         0.172   0.358    9  ...     150.857               4      10
41095         0.910   0.366    1  ...     152.000               4      10
41096         0.719   0.804   10  ...     227.760               4      10
41097         0.600   0.177    7  ...     213.387               4      10
41098         0.121   0.123    4  ...     341.396               4      10

[40750 rows x 14 columns]

### &emsp;&emsp; 3.1.2 Preprocessing
Before running our models, our data should be preprocessed to increase performance. In this case, we would be modifying all of our numerical features to be on the same scale (values between 0 and 1). This is for two main reasons:
<br> 1) Our features have significantly different ranges. This can be problematic because it can lead to several features dominating the algorithm.
<br> 2) Our features are not measured in the same units, so our model may be unit-dependent and thus reliant on how the feature is measured.

In some ways, preprocessing greatly reduces the risk of the above problems. We split the dataset and performed preprocessing below. 

In [ ]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

#split the dataset
def split_the_dataset():
    return train_test_split(features, target, random_state=3000)

X_train, X_test, y_train, y_test = split_the_dataset() #store the variables after splitting

#create the scaler, fit it with the training data, and scale the feature variables
scaler = MinMaxScaler()
scaler.fit(X_train) 
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
X_train_scaled

array([[0.17884781, 0.49287271, 0.81818182, ..., 0.11256751, 0.6       ,
        0.11111111],
       [0.6216681 , 0.25581321, 0.09090909, ..., 0.0237953 , 0.8       ,
        0.66666667],
       [0.638865  , 0.45686367, 0.90909091, ..., 0.03919137, 0.8       ,
        0.66666667],
       ...,
       [0.37016337, 0.04576049, 0.45454545, ..., 0.03460192, 0.8       ,
        1.        ],
       [0.27450559, 0.57789405, 0.90909091, ..., 0.09987707, 0.8       ,
        0.77777778],
       [0.68400688, 0.46686618, 0.81818182, ..., 0.06590437, 0.8       ,
        1.        ]])

In [ ]:
X_test_scaled

array([[0.31319862, 0.92998243, 0.63636364, ..., 0.03823502, 0.8       ,
        0.77777778],
       [0.24763543, 0.19079689, 0.90909091, ..., 0.0546197 , 0.8       ,
        0.66666667],
       [0.61521926, 0.66091489, 1.        , ..., 0.02855831, 0.8       ,
        0.66666667],
       ...,
       [0.4572227 , 0.25981421, 0.81818182, ..., 0.06188276, 0.8       ,
        0.88888889],
       [0.17454858, 0.20580066, 0.        , ..., 0.03586952, 0.6       ,
        0.66666667],
       [0.37983663, 0.78794677, 0.27272727, ..., 0.05489243, 0.6       ,
        0.77777778]])

### &emsp;&emsp; 3.1.3 Feature Extraction

One-hot encoding is used when the machine learning algorithm needs to determine the difference between categorical variables. This entails creating new features from existing features and using 0 or 1 to represent if the data belongs/does not belong to a certain class.  

The only string variable we have is the 'target,' but because that is the target variable, one-hot encoding is not applicable to our dataset. Thus, we could not apply any feature extraction techniques. 


### &emsp;&emsp; 3.1.4 Feature Selection

Because not all features may be equally important in predicting the music genres, we use feature extraction to get the most important ones. Doing so can increase accuracy and algorithm speed for large datasets. We have chosen to use Recursive Feature Elimination (RFE), which starts with all the features, builds the model, eliminates the least important feature, and repeats the process. Because of its recursive nature, RFE is computationally expensive compared to other feature elimination techniques. Nevertheless, RFE often leads to better results compared to other feature elimination techniques because of this.

In [ ]:
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeRegressor

#use rfe and the decision tree regressor to decide whether or not to keep the feature or not. We chose to only use the 5 most important features
select = RFE(DecisionTreeRegressor(random_state = 3000), n_features_to_select=5)
select.fit(X_train_scaled, y_train)

#scale only the selected features based on previous preprocessing
X_train_selected = select.transform(X_train_scaled)
X_test_selected = select.transform(X_test_scaled)

#print results of RFE
print('Selected features after RFE:')
bool_list = select.get_support()
for i in range(len(bool_list)):
    if (bool_list[i]): #if "True":
        print('\t' + features.columns[i])

Selected features after RFE:
	danceability
	energy
	speechiness
	tempo
	duration_s


In [ ]:
X_train_selected

array([[0.17884781, 0.49287271, 0.01044776, 0.52865715, 0.11256751],
       [0.6216681 , 0.25581321, 0.02377399, 0.46202653, 0.0237953 ],
       [0.638865  , 0.45686367, 0.00554371, 0.21639041, 0.03919137],
       ...,
       [0.37016337, 0.04576049, 0.01172708, 0.34446954, 0.03460192],
       [0.27450559, 0.57789405, 0.01481876, 0.53458572, 0.09987707],
       [0.68400688, 0.46686618, 0.01162047, 0.52489617, 0.06590437]])

In [ ]:
X_test_selected

array([[0.31319862, 0.92998243, 0.04701493, 0.7265154 , 0.03823502],
       [0.24763543, 0.19079689, 0.0196162 , 0.16570808, 0.0546197 ],
       [0.61521926, 0.66091489, 0.00458422, 0.45237047, 0.02855831],
       ...,
       [0.4572227 , 0.25981421, 0.0315565 , 0.46361992, 0.06188276],
       [0.17454858, 0.20580066, 0.01119403, 0.55342412, 0.03586952],
       [0.37983663, 0.78794677, 0.21855011, 0.648932  , 0.05489243]])

### 3.2. Data Exploration
In order to analyze and better understand the song data that we are working with, we created various visualizations that show the relationships between the different features and genres. All images can be found in the GitHub repo [here.](https://github.com/edshen17/FP4)

First, we import plotly.express in order to use some of the visualizations that it offers.

In [ ]:
import plotly.express as px

In [ ]:
genre_counts = df_original["genre"].value_counts()
count_fig = px.bar(x=genre_counts.index, y=genre_counts, color=genre_counts.index, labels={"x":"Genre", "y":"Number of Songs"}, title="Count by Genre")
count_fig.show()

<img src="https://raw.githubusercontent.com/edshen17/FP4/master/bar.png"/>



When working with grouped data, there could be classes that do not contain many values in the training/testing data. If there was, for example, one genre with only a few songs, this could negatively influence the accuracy of the classification. However, as shown by the bar plot above, this is not the case for our data. For each genre, which is represented by the x-axis of the graph, there are at least 1500 songs, as shown by the y-axis of the graph.

As a result, we can be confident that our algorithm and subsequent work will allow us to determine if there is an association between our features and the genres of songs. We can see that while pop music is the most frequent genre in our data, edm is the least frequent.


In [ ]:
scatter_genre = px.scatter_matrix(df_original, dimensions= ["danceability", "speechiness", "tempo"], color="genre", title="Feature Scatter Matrix")
scatter_genre.show()

<img src="https://raw.githubusercontent.com/edshen17/FP4/master/scatter.png"/>

Next, we created a scatter matrix. We used this to compare the 3 most important features obtained through our feature selection process. These scatter plots have either danceablility, speechiness, or tempo on the x and y-axes. Additionally, the data points are colored based on genre.

While not inherently indicative of associations, scatter matrices can provide important insights into the relationships between various features in the data. The results of this scatter matrix show that the genres are distinct in regards to the features given, leading us to believe that we will likely be able to accurately classify songs based on this data.

In [ ]:
scatter3d_genre = px.scatter_3d(df_original, x="danceability", y="speechiness", z="tempo", color="genre", title="Three-Dimensional Feature Plot")
scatter3d_genre.show()

<img src="https://raw.githubusercontent.com/edshen17/FP4/master/3d.png"/>

To further explore and expand upon the data from the above scatter matrix, we created a 3-dimensional scatter plot. This time, we set speechiness, danceability, and tempo to be the x, y, and z-axes for the scatter plot. Again, we chose these features because they were the 3 most important ones after feature selection. Like before, we set the color of each data point to represent the genre to which that song belongs.

This 3-dimensional plot gave us important insights into the data. It seems like there are some solid differences between genre as there is lots of variation. It also seems that songs in the rap, rock, and EDM genres might be easier to accurately classify because they are quite distinct and separate from other classes. However, the other genres are very close to each other and leads us to believe that we need more than 3 selected features.

NOTE: We have talked to the professor about this visualization and whether it is valid for extra credit, as it is technically a "scatterplot." However, he said he would make the exception for this one as we applied a new visualization not learned in class.

### 3.3. Model Construction

### &emsp;&emsp; 3.3.1. Hypothesis Testing


In [ ]:
#function to perform statistical analysis
def t_test(data, IV1, IV2):
    from scipy import stats

    #run Welch's t-test
    t_stat, p_val = stats.ttest_ind(df[IV1], df[IV2], equal_var=False)

    #run a pearson test to get the pearson coefficient
    r, p = stats.pearsonr(df[IV1], df[IV2])

    #get degrees of freedom
    deg_free = len(df[IV1]) + len(df[IV2]) - 2

    #print out the stats
    print("t(%d)=%.2f, p=%.3f" % (deg_free, t_stat, p_val))
    print("Pearson coefficient:", str(r) + ",", "p=", p)

In [ ]:
t_test(df, "danceability", "energy")

t(81498)=-25.95, p=0.000
Pearson coefficient: 0.2043393284539016, p= 0.0


### &emsp;&emsp; 3.3.2 Training/Test Split

The split was performed in section 3.1.2 (Preprocessing)

### &emsp;&emsp; 3.3.3 Applying ML Algorithms

In [ ]:
for classifier in estimators: #iterate through our estimator object and use the classifiers
        estimators[classifier].fit(X=X_train_selected, y=y_train) #fit the model to the data/selected features
        train_accuracy = estimators[classifier].score(X_train_selected, y_train) #get accuracy of our model on training set
        accuracy = estimators[classifier].score(X_test_selected, y_test) #get accuracy of our model on test set
        print(classifier)
        print('\tPrediction accuracy on the training data: ' + format(train_accuracy*100, '.2f') + '%\n')
        print('\tPrediction accuracy on the test data: ' + format(accuracy*100, '.2f') + '%\n')

k-Nearest Neighbor
	Prediction accuracy on the training data: 90.96%

	Prediction accuracy on the test data: 85.04%

Support Vector Machine
	Prediction accuracy on the training data: 87.12%

	Prediction accuracy on the test data: 87.31%

Decision Tree
	Prediction accuracy on the training data: 100.00%

	Prediction accuracy on the test data: 99.97%



### 3.4. Model Evaluation


### &emsp;&emsp; 3.4.1. Model Accuracy

Accuracy is an indication of how well the estimator performs on classifying the data correctly. For example, if we had a 90% accuracy from our algorithm, that means that we correctly classified the songs' genre 90% of the time. Likewise, this also means we incorrectly classified the songs 10% of time (eg. error rate of 10%). As seen in section 3.3.3, we checked the accuracy score of each model after fitting them to the selected set. 

Using the 5 selected features, we obtained 85.04% accuracy for the k-nearest neighbor, 87.31% for the SVM algorithm, and 99.97% for the decision tree algorithm on the test sets. The first 2 algorithms have solid accuracies and show no sign of overfitting. However, the decision tree's perfect accuracy on training set indicates that the model is overfitting. Still, from the results above, we believe that the decision tree looks the most promising for solving our problem as it yields the highest results.

Note: For other evaluation metrics, refer to section 3.4.3 (classification report).

### &emsp;&emsp; 3.4.1.1 Hypothesis Testing Results
* An independent sample t-test was conducted to compare a song's danceability and its energy. Results showed a statistically significant correlation between the two features,  t(81498)=-25.95, p=0.000, and R=.204. As p < .001, these results indicate that there is a strong correlation between danceability values and energy values in a song. This refutes our original hypothesis that there is no correlation between danceability and energy.

* The accuracy and mean standard error were examined for the SVM, KNN, and Decision Tree models. Results from sections 3.3.3, 3.4.2, and 3.4.3 show that the decision tree has better accuracy, precision, recall, and F1 compared to the other algorithms. This indicates that it is better for classifying and predicting song genres. Thus, these results refute our original hypothesis that KNN and SVM would be better at classifying song genres.


### &emsp;&emsp; 3.4.2. Cross Validation

We ran cross-validation on each of our algorithms to double-check their accuracy. The mean values for the cross-validation scores were slightly lower, but the results are similar enough to confirm that the k-nearest neighbor and support vector machine algorithms are less accurate than the decision tree algorithm. We also did this to see whether the decision tree was being overfit.

In [ ]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

kfold = KFold(n_splits=10, random_state=3000, shuffle=True)

for classifier in estimators:
  #get the scores after dividing the data into 10 splits/cross validation
  scores = cross_val_score(estimator=estimators[classifier], X=X_train_selected, y=y_train, cv=kfold)
  print(classifier)
  print(f'\tMean accuracy: {scores.mean():.2%}')
  print(f'\tStandard deviation={scores.std():.2%}')
  

k-Nearest Neighbor
	Mean accuracy: 84.09%
	Standard deviation=0.38%
Support Vector Machine
	Mean accuracy: 86.93%
	Standard deviation=0.50%
Decision Tree
	Mean accuracy: 99.99%
	Standard deviation=0.02%


### &emsp;&emsp; 3.4.3. Classification Report

Additionally, we decided to create a classification report. Results support our conclusions from before.

The metrics of this classification report are:
* **Precision**: The proportion of correctly predicted positive classes to all positive predictions.
* **Recall**: The proportion of correctly predicted positive classes to all positive classes. In other words, recall is the true positive rate or the sensitivity/probability of detection.
* **F-1 score**: a weighted harmonic mean of the precision and recall, where an F-1 score reaches its best value at 1 and worst score at 0.
* **Support**: the number of occurrences of each class

In [ ]:
from sklearn.metrics import classification_report

for classifier in estimators: #iterate through our estimator object and use the classifiers
        estimators[classifier].fit(X=X_train_selected, y=y_train) #fit the model to the data/selected features
        class_report = classification_report(predicted, y_test) # get the classification report
        print(classifier)
        print(class_report)

k-Nearest Neighbor
              precision    recall  f1-score   support

           1       1.00      1.00      1.00      4163
           2       1.00      1.00      1.00      2676
           3       1.00      1.00      1.00      1678
           4       1.00      1.00      1.00       724
           5       1.00      1.00      1.00       602
           6       1.00      1.00      1.00       345

    accuracy                           1.00     10188
   macro avg       1.00      1.00      1.00     10188
weighted avg       1.00      1.00      1.00     10188

Support Vector Machine
              precision    recall  f1-score   support

           1       1.00      1.00      1.00      4163
           2       1.00      1.00      1.00      2676
           3       1.00      1.00      1.00      1678
           4       1.00      1.00      1.00       724
           5       1.00      1.00      1.00       602
           6       1.00      1.00      1.00       345

    accuracy                       

For convenience, we have included the genre_dict below (the dictionary includes the string name of each numerical class).

In [ ]:
genre_dict

{1: 'pop', 2: 'r&b', 3: 'rock', 4: 'latin', 5: 'rap', 6: 'edm'}

We can see that the decision tree is outperforming all the other algorithms in terms of precision, recall, and f1-score. We also see that overall, kNN and SVM perform similarly in terms of these metrics, with SVM doing a little better on average.

### 3.5. Model Optimization

We next chose to tune each of our models in order to determine if we can improve their accuracy. Additionally, our 100% accuracy on the training set for the decision tree model indicates that it is very likely to be overfitting the data. Tuning can be used to prevent this overfitting and improve the model's ability to generalize. We use Scikit-learn's GridSearchCV method to test various parameters and determine the best ones for each algorithm. 

Note: As we couldn't tell beforehand what hypertuning would do to our decision tree algorithm, we decided to tune all the algorithms just in case.

In [ ]:
from sklearn.model_selection import GridSearchCV

### &emsp;&emsp; 3.5.1. K-Nearest Neighbor

For the k-nearest neighbors algorithm, we can optimize the n_neighbors and metric parameters. As we recognize that lower n_neighbors values have the potential to overfit the data, we tried out a variety of values. We also tried 3 different distance functions (metric) to see if changing them would lead to better results.

In [ ]:
#create dictionary with all the possible tuning values
knn_param_grid = {"n_neighbors":[3, 5, 7, 9], "metric": ["euclidean", "manhattan", "minkowski"]}

grid_search_knn = GridSearchCV(KNeighborsClassifier(), knn_param_grid)

grid_search_knn.fit(X=X_train_selected, y=y_train)

# get the best values
best_n_neighbors = grid_search_knn.best_params_["n_neighbors"]
best_k_metric = grid_search_knn.best_params_["metric"]

print("K-Nearest Neighbor Classifier")
print("Best parameters: ", grid_search_knn.best_params_)
print("Best cross-validation score: ", grid_search_knn.best_score_)

print("Test set score: ", grid_search_knn.score(X_test_selected, y_test))

K-Nearest Neighbor Classifier
Best parameters:  {'metric': 'manhattan', 'n_neighbors': 9}
Best cross-validation score:  0.8721287587263603
Test set score:  0.8806438947781704


The results seem to indicate that the ideal value for n_neighbors is 9, which yields a slight increase in the cross-valdation and test set scores.

### &emsp;&emsp; 3.5.2. Support Vector Machine

Although the results from 3.3 and 3.4 did not indicate that the SVM model was overfitting, we decided to see if we could improve its accuracy. We used GridSearchCV to check whether the default C value is ideal.

In [ ]:
#create dictionary with all the possible tuning values
svm_param_grid = {"C":[.01, .1, 1]}

grid_search_svm = GridSearchCV(LinearSVC(max_iter=100000), svm_param_grid)

grid_search_svm.fit(X=X_train_selected, y=y_train)
best_C = grid_search_svm.best_params_["C"]

#result of grid search
print("Best parameters: ", grid_search_svm.best_params_)
print("Best cross-validation score: ", grid_search_svm.best_score_)

#the performance of the best found parameters on the test set
#this is what you report for the evaluation of your model
print("Test set score: ", grid_search_svm.score(X_test_selected, y_test))

Best parameters:  {'C': 1}
Best cross-validation score:  0.8683006154594578
Test set score:  0.8730859835100118


As expected, the parameter C with a value of 1 performed the best. Our SVM model, though not overwhelmingly accurate, performs well and appears to not be overfitting the data.

### &emsp;&emsp; 3.5.3. Decision Tree

Next, we tuned the decision tree model. This model almost certainly is overfitting the data with the default parameters, so we tuned the max_depth and max_features parameters.

In [ ]:
#create dictionary with all the possible tuning values
tree_param_grid = {"max_depth": [2, 3, 4, 5], "max_features": ["auto", "sqrt", "log2"]}

grid_search_tree = GridSearchCV(DecisionTreeClassifier(), tree_param_grid)

grid_search_tree.fit(X=X_train_selected, y=y_train)
best_max_depth = grid_search_tree.best_params_["max_depth"]
best_max_features= grid_search_tree.best_params_["max_features"]

#result of grid search
print("Best parameters: ", grid_search_tree.best_params_)
print("Best cross-validation score: ", grid_search_tree.best_score_)

#the performance of the best found parameters on the test set
print("Test set score: ", grid_search_tree.score(X_test_selected, y_test))

Best parameters:  {'max_depth': 5, 'max_features': 'auto'}
Best cross-validation score:  0.9406133225646485
Test set score:  0.9268747546132705


### &emsp;&emsp; 3.5.4. Updating the Models

Lastly, we updated the models with these new hyperparameters, as shown below. Although the sections above show the results, we do this just for completion.

In [ ]:
estimators["k-Nearest Neighbor"] = KNeighborsClassifier(n_neighbors=best_n_neighbors, metric=best_k_metric)
estimators["Support Vector Machine"] = LinearSVC(C=best_C, random_state=3000)
estimators["Decision Tree"] = DecisionTreeClassifier(max_depth=best_max_depth, max_features=best_max_features, random_state=3000)

### 3.6. Model Testing
Finally, we tested our models again with the testing set. We do this to be able to more clearly compare the algorithms, as all the data would be printed in one place.

In [ ]:
for classifier in estimators: #iterate through our estimator object and use the classifiers
        estimators[classifier].fit(X=X_train_selected, y=y_train) #fit the model to the data/selected features
        accuracy = estimators[classifier].score(X_test_selected, y_test) #get accuracy of our model on test set
        print(classifier)
        print('\tPrediction accuracy on the test data: ' + format(accuracy*100, '.2f') + '%\n')

k-Nearest Neighbor
	Prediction accuracy on the test data: 88.06%

Support Vector Machine
	Prediction accuracy on the test data: 87.31%

Decision Tree
	Prediction accuracy on the test data: 96.72%



<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION
Through this project, we compared the k-nearest neighbor algorithm, the support vector machine algorithm, and the decision tree algorithm. The decision tree algorithm by far performed the best, but also required hyperparameter tuning to prevent overfitting. Because of this tuning, however, we are confident in our decision tree model's ability to generalize, and believe that this algorithm should be used for our predictive model. As all 3 of our algorithms yielded results with accuracy higher than 85% for the test set, we have confidence that it is possible to use song features from Spotify to accurately classify that song to its genre.

Through hypothesis testing, we were able to identify significant relationships in our data that was unknown previously. We saw that there is a correlation between energy and danceability of a song, which suggests that people are more likely to dance to a song when it is very high beat and energetic. In some ways, these results could help others choose songs when they want people to dance or not.

For the future, a potential project that builds on the models we created is a website that allows the user to input a song from Spotify. Then, given the song, the website outputs the predicted genre of that song based on its features. Other potential future project could be to gather the features of a song based on its audio rather than obtaining them from Spotify. This would allow a person to record a song and get its genre. We could also scrape our own data and have more genres to classify, as we only had 6. While there is a lot of room to expand on the work we have done in this project, our project itself demonstrates the application of machine learning in the field of music.



<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

### CONTRIBUTIONS

The majority of the project was completed as a group over Microsoft Teams. We did this to ensure all members would have a solid understanding of the material. However, there were some aspects of the project that each member played a larger role in completing:


*   Edwin Shen - Data Wrangling, Model Training 
*   Tatum Gaudinski Whitehead - Hypothesis Creation and Testing, Significance of Problem, Synthesis of Previous Work on Subject
*   Benjamin Wolff - Data Exploration, Hyperparameter Tuning



References: 
* (1) Middlebrook, K., & Sheik, K. (2019). Song Hit Prediction: Predicting Billboard Hits Using Spotify Data. Cornell University. https://arxiv.org/abs/1908.08609 
* (2) Greer, T., & Narayanan, S. (2019). Using Shared Vector Representations of Words 
     and Chords in Music for Genre Classification. https://www.isca-speech.org/archive/SMM_2019/pdfs/SMM19_paper_19.pdf 
* (3) Pelchat, N., & Gelowitz, C. M. (2019). Neural Network Music Genre 
     Classification. IEEE. https://ieeexplore.ieee.org/document/8861555 
* (4) Spotify Genre Joined Dataset. https://www.kaggle.com/akiboy96/spotify-genre-joined
* (5) Spotify - Get Audio Features for a Track. https://developer.spotify.com/documentation/web-api/reference/tracks/get-audio-features/
